clang/docs/checkedc/Bounds-Declaration-Checking.md

```diff
@@ -0,0 +1,206 @@
+# Bounds Declaration Checking
+
+## Checking Overview
+After each statement in the clang CFG, the inferred bounds for an expression
+must imply the expression's declared bounds. The algorithms for inferring and
+checking bounds for expressions are implemented in [SemaBounds.cpp](https://github.com/microsoft/checkedc-clang/blob/master/clang/lib/Sema/SemaBounds.cpp).
+
+## Bounds Terminology
+The bounds checking code uses the following definitions to reason about
+bounds expressions:
+- **Declared bounds**: The bounds that the programmer has declared for a
+pointer expression.
+  -Example: In `array_ptr<int> p : bounds(p, p + 1) = 0;`, the declared bounds
+  of `p` are `bounds(p, p + 1)`.
+- **Inferred bounds**: The bounds that the compiler determines for a pointer
+expression at a particular point in checking a statement.
+  -Example: After checking `array_ptr<int> p : bounds(p, p + 1) = 0`, the
+  inferred bounds of `p` are `bounds(any)` (since `0` by definition has
+  bounds of `bounds(any)`).
+- **RValue bounds**: The bounds of the value produced by an rvalue expression.
+  - Example: If `p` is a variable with declared bounds of `bounds(p, p + 1)`,
+  then the rvalue expression `p + 3` has rvalue bounds of `bounds(p, p + 1)`.
+  - Example: If `p` is a variable with declared bounds of `bounds(p, p + 5)`,
+  then the rvalue expression from reading the value of `*(p + 2)` has rvalue
+  bounds of `bounds(unknown)`.
+- **LValue bounds**: The bounds of an lvalue expression `e`. These bounds
+determine whether it is valid to access memory using `e`, and should be the
+range (or a subrange) of an object in memory. LValue bounds are used to check
+that a memory access is within bounds.
+  - Example: If `p` is a non-array-typed variable, then the lvalue bounds
+  of `p` are `bounds(&p, &p + 1)`.
+  - Example: If `p` is a variable with declared bounds of `bounds(p, p + 3)`,
+  then `*(p + 1)` and `p[2]` have lvalue bounds of `bounds(p, p + 3)`. The
+  compiler will check that `p + 1` and `p + 2` are within the lvalue bounds of
+  `bounds(p, p + 3)`.
+- **Target bounds**: The target bounds of an lvalue expression `e`. Values
+assigned to `e` must satisfy these bounds. Values read through `e` will
```

Comment:
Following up on our earlier offline discussion: the target bounds of a dereference expression `*e1` are `bounds(unknown)` (unless `e1` is a `_Ptr` or `_Nt_checked[]` type). If `p` has bounds `(p, p + 100)` and `q` has bounds `(q, q + 3)`, the assignment `*(p + 5) = q` has LHS target bounds of `bounds(unknown)`, so the assignment is valid.